gobblin-yarn/src/main/java/org/apache/gobblin/yarn/YarnService.java

```diff
@@ -339,6 +342,15 @@ protected void startUp() throws Exception {
     LOGGER.info("ApplicationMaster registration response: " + response);
     this.maxResourceCapacity = Optional.of(response.getMaximumResourceCapability());
 
+    // All previous helix instances should be purged on startup. Gobblin task runners are stateless from helix
+    // perspective because all important state is persisted separately in Workunit State Store or Watermark store.
+    // Offline duration of 0 means any offline instance should be purged (Note: there aren't any online instances
+    // when this code runs, this is during startup before any containers are allocated).
+    LOGGER.info("Purging offline helix instances before allocating containers for helixClusterName={}, connectionString={}", helixManager.getClusterName(), helixManager.getMetadataStoreConnectionString());
+    long offlineDuration = 0;
+    this.helixAdmin.purgeOfflineInstances(this.helixManager.getClusterName(), offlineDuration);
```

Comment:
Not sure if we should have some sort of timeout in case the API takes a long time to respond. Currently this a blocking call and I wonder if this call fails we could block containers from being allocated in the YarnService (leading to starvation).

But adding a timeout adds a potential bug / risk where we timeout and then helix starts purging instances while we are allocating new instances (bad bug with nondeterministic behavior)

Reply:
Well kind of. As mentioned

> But adding a timeout adds a potential bug / risk where we timeout and then helix starts purging instances while we are allocating new instances (bad bug with nondeterministic behavior)

We put ourselves at risk of having a live instances without an INSTANCE config. That would essentially bork the live instance and have a domino effect. Although I do think we should send a GTE if this call is taking too long

Reply:
Agreed that we need to avoid the race condition, according to [ZKHelixAdmin.purgeOfflineInstances()](https://github.com/apache/helix/blob/ee61ff434ee37d4ab8456c739b2229f99d9e0e72/helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java#L270). 
Please do add a GTE or an alert here if this takes too long

Reply:
We should also add a config here to make this action configurable

Reply:
Thanks for linking the code documentation. Forgot to link that

Reply:
+1 to make this behavior configurable, so that if something wrong happens on helix side and causing the starvation we can quickly mitigate by disable this behavior. Also it will be good to add instrument here to measure how long it usually take for us to clear the offline instance.